[SPARK-27270][SS] Add Kafka dynamic JAAS authentication debug possibility

gaborgsomogyi opened the pull request:
## What changes were proposed in this pull request?

`Krb5LoginModule` supports debug parameter which is not yet supported from Spark side. This configuration makes it easier to debug authentication issues against Kafka.

In this PR `Krb5LoginModule` debug flag controlled by either `sun.security.krb5.debug` or `com.ibm.security.krb5.Krb5Debug`.

Additionally found some hardcoded values like `ssl.truststore.location`, etc... which could be error prone if Kafka changes it so in such cases Kafka define used.

## How was this patch tested?

Existing + additional unit tests + on cluster.

gaborgsomogyi committed:
[SPARK-27270][SS] Add Kafka dynamic JAAS authentication debug possibility

gaborgsomogyi commented:
cc @HeartSaVioR

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #103915 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103915/testReport)** for PR 24204 at commit [`bc210c6`](https://github.com/apache/spark/commit/bc210c6edec1358167aed0c8ccf69d7d9d2724e4).

SparkQA commented:
**[Test build #103915 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103915/testReport)** for PR 24204 at commit [`bc210c6`](https://github.com/apache/spark/commit/bc210c6edec1358167aed0c8ccf69d7d9d2724e4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103915/
Test PASSed.

HeartSaVioR reviewed (approved):
LGTM

vanzin commented:
Is this different from setting the `sun.security.krb5.debug` system property? That doesn't require any changes to Spark.

HeartSaVioR commented:
I'm seeing some value to enable the option to only Kafka side instead of global (since that option normally brings verbose messages), but no strong opinion.

gaborgsomogyi commented:
It provides different kind of debug log entries just like `HADOOP_JAAS_DEBUG`. Please see [doc](https://www.cloudera.com/documentation/enterprise/5-9-x/topics/sg_kerberos_troubleshoot.html#concept_ngz_wjd_sw) (I know its CM but not found better) and the [code](https://github.com/apache/hadoop/blob/a55d6bba71c81c1c4e9d8cd11f55c78f10a548b0/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/delegation/ZKDelegationTokenSecretManager.java#L270).

> I'm seeing some value to enable the option to only Kafka side instead of global (since that option normally brings verbose messages), but no strong opinion.

My change turns on debug messages only on Kafka related authentications. Not really sure what flag do you mean.

HeartSaVioR commented:
> My change turns on debug messages only on Kafka related authentications. Not really sure what flag do you mean.

That's what I meant in my comment, vs turning global configuration.

vanzin commented:
> It provides different kind of debug log entries just like HADOOP_JAAS_DEBUG

I didn't ask about that option. I asked about `sun.security.krb5.debug` which is the system property to enable the exact thing you're enabling here with another config.

The only advantage I see here is to only enable that option for the kafka login, but is that really an advantage that justifies adding a new config option that people won't necessarily know about, when everybody who has to debug kerberos one time or another has to deal with that system property?

Did you try setting that system property and seeing whether it shows the things you want to see when debugging?

dongjoon-hyun commented:
I also agree with @vanzin 's point of views. For the security stuff, we need to see all logs in many cases. I believe the current documentation is enough.
- https://spark.apache.org/docs/latest/running-on-yarn.html#troubleshooting-kerberos

gaborgsomogyi commented:
Maybe my explanation was not enough/clean. Let me give a little bit more details.
I've created a small standalone application where these things can be tested easily.
The application connect to a secure Kafka cluster and tries to do authentication with dynamic JAAS configuration (where `debug=true` can be set). I've re-tested everything to give exact logs.

Test with `debug=false`:
Command: `java -Dsun.security.krb5.debug=true -Dsun.security.spnego.debug=true ...`
Output:
```
Java config name: null
Native config name: /etc/krb5.conf
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 70; type: 16
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 62; type: 23
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 54; type: 8
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 54; type: 3
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 54; type: 1
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 54; type: 1
Looking for keys for: systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (3) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (8) for systest@GCE.CLOUDERA.COM
Added key: 23version: 1
Added key: 16version: 1
>>> KdcAccessibility: reset
Looking for keys for: systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (3) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (8) for systest@GCE.CLOUDERA.COM
Added key: 23version: 1
Added key: 16version: 1
Using builtin default etypes for default_tkt_enctypes
default etypes for default_tkt_enctypes: 17 16 23.
>>> KrbAsReq creating message
getKDCFromDNS using UDP
getKDCFromDNS using TCP
>>> KrbKdcReq send: kdc=gsomogyi-cdh6x-secure-1.gce.cloudera.com UDP:88, timeout=30000, number of retries =3, #bytes=148
>>> KDCCommunication: kdc=gsomogyi-cdh6x-secure-1.gce.cloudera.com UDP:88, timeout=30000,Attempt =1, #bytes=148
>>> KrbKdcReq send: #bytes read=775
>>> KdcAccessibility: remove gsomogyi-cdh6x-secure-1.gce.cloudera.com
Looking for keys for: systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (3) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (8) for systest@GCE.CLOUDERA.COM
Added key: 23version: 1
Added key: 16version: 1
>>> EType: sun.security.krb5.internal.crypto.Des3CbcHmacSha1KdEType
>>> KrbAsRep cons in KrbAsReq.getReply systest
>>> 19/04/01 10:08:56 INFO authenticator.AbstractLogin: Successfully logged in.
>>> 19/04/01 10:08:56 INFO kerberos.KerberosLogin: [Principal=systest@GCE.CLOUDERA.COM]: TGT refresh thread started.
>>> 19/04/01 10:08:56 INFO kerberos.KerberosLogin: [Principal=systest@GCE.CLOUDERA.COM]: TGT valid starting at: Mon Apr 01 10:08:55 CEST 2019
>>> 19/04/01 10:08:56 INFO kerberos.KerberosLogin: [Principal=systest@GCE.CLOUDERA.COM]: TGT expires: Wed May 01 10:08:55 CEST 2019
>>> 19/04/01 10:08:56 INFO kerberos.KerberosLogin: [Principal=systest@GCE.CLOUDERA.COM]: TGT refresh sleeping until: Fri Apr 26 05:03:45 CEST 2019
>>> 19/04/01 10:08:56 INFO utils.AppInfoParser: Kafka version : 2.0.0
>>> 19/04/01 10:08:56 INFO utils.AppInfoParser: Kafka commitId : 3402a8361b734732
```

Test with `debug=true`:
Command: `java ...`
Output:
```
Java config name: null
Native config name: /etc/krb5.conf
Debug is  true storeKey true useTicketCache false useKeyTab true doNotPrompt false ticketCache is null isInitiator true KeyTab is /Users/gaborsomogyi/kafka-delegation-token/kafka-consumer/systest.keytab refreshKrb5Config is false principal is systest@GCE.CLOUDERA.COM tryFirstPass is false useFirstPass is false storePass is false clearPass is false
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 70; type: 16
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 62; type: 23
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 54; type: 8
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 54; type: 3
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 54; type: 1
>>> KeyTabInputStream, readName(): GCE.CLOUDERA.COM
>>> KeyTabInputStream, readName(): systest
>>> KeyTab: load() entry length: 54; type: 1
Looking for keys for: systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (3) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (8) for systest@GCE.CLOUDERA.COM
Added key: 23version: 1
Added key: 16version: 1
>>> KdcAccessibility: reset
Looking for keys for: systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (3) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (8) for systest@GCE.CLOUDERA.COM
Added key: 23version: 1
Added key: 16version: 1
Using builtin default etypes for default_tkt_enctypes
default etypes for default_tkt_enctypes: 17 16 23.
>>> KrbAsReq creating message
getKDCFromDNS using UDP
getKDCFromDNS using TCP
>>> KrbKdcReq send: kdc=gsomogyi-cdh6x-secure-1.gce.cloudera.com UDP:88, timeout=30000, number of retries =3, #bytes=148
>>> KDCCommunication: kdc=gsomogyi-cdh6x-secure-1.gce.cloudera.com UDP:88, timeout=30000,Attempt =1, #bytes=148
>>> KrbKdcReq send: #bytes read=775
>>> KdcAccessibility: remove gsomogyi-cdh6x-secure-1.gce.cloudera.com
Looking for keys for: systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (1) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (3) for systest@GCE.CLOUDERA.COM
Found unsupported keytype (8) for systest@GCE.CLOUDERA.COM
Added key: 23version: 1
Added key: 16version: 1
>>> EType: sun.security.krb5.internal.crypto.Des3CbcHmacSha1KdEType
>>> KrbAsRep cons in KrbAsReq.getReply systest
principal is systest@GCE.CLOUDERA.COM
Will use keytab
Commit Succeeded

>>> 19/04/01 10:11:25 INFO authenticator.AbstractLogin: Successfully logged in.
>>> 19/04/01 10:11:25 INFO kerberos.KerberosLogin: [Principal=systest@GCE.CLOUDERA.COM]: TGT refresh thread started.
>>> 19/04/01 10:11:25 INFO kerberos.KerberosLogin: [Principal=systest@GCE.CLOUDERA.COM]: TGT valid starting at: Mon Apr 01 10:11:24 CEST 2019
>>> 19/04/01 10:11:25 INFO kerberos.KerberosLogin: [Principal=systest@GCE.CLOUDERA.COM]: TGT expires: Wed May 01 10:11:24 CEST 2019
>>> 19/04/01 10:11:25 INFO kerberos.KerberosLogin: [Principal=systest@GCE.CLOUDERA.COM]: TGT refresh sleeping until: Fri Apr 26 02:37:27 CEST 2019
>>> 19/04/01 10:11:25 INFO utils.AppInfoParser: Kafka version : 2.0.0
>>> 19/04/01 10:11:25 INFO utils.AppInfoParser: Kafka commitId : 3402a8361b734732
```

As a final conclusion even if global krb debug flag is enabled `Krb5LoginModule` debug messages not shown. Please see from the first execution the following message is missing:
```
Debug is  true storeKey true useTicketCache false useKeyTab true doNotPrompt false ticketCache is null isInitiator true KeyTab is /Users/gaborsomogyi/kafka-delegation-token/kafka-consumer/systest.keytab refreshKrb5Config is false principal is systest@GCE.CLOUDERA.COM tryFirstPass is false useFirstPass is false storePass is false clearPass is false
...
principal is systest@GCE.CLOUDERA.COM
Will use keytab
Commit Succeeded
...
```
These messages is just a sample and `Krb5LoginModule` provides much more debug information which may be helpful for debugging.

gaborgsomogyi commented:
To make life easier and testable by others as well the app can be found [here](https://github.com/gaborgsomogyi/kafka-delegation-token).

vanzin commented:
> Please see from the first execution the following messages are missing:

I guess I'm not seeing what's so useful about those messages. All the other stuff that is printed when you set the system property seems way more useful.

gaborgsomogyi commented:
That said the module contains 22 debug messages not only the 2 shown up.
The advantages what I see:
* It provides messages from higher level of abstraction
* Prints out exceptions for example [here](https://github.com/tinyking/jdk1.8/blob/1bc0a93318e8986f56fb1b7f077fc6e6dca92d6e/src/com/sun/security/auth/module/Krb5LoginModule.java#L626)

gaborgsomogyi commented:
If there are still concerns maybe the debug flag can be turned on by `sun.security.krb5.debug` which will end up in not introducing new Spark config.

vanzin commented:
If you really think those extra debug messages are super important, I'd reuse the known system property for that purpose.

vanzin commented:
> Prints out exceptions for example here

The exception is also thrown, so it would show up anyway.

gaborgsomogyi committed:
Switch to global debug flag

gaborgsomogyi commented:
After dealing several weeks with multi-cluster delegation token and facing constant kerberos problems I have 2 conclusions:
* I see your point and from user perspective having a new flag is overkill and not convenient at all
* Kerberos core provided messages are sometimes cryptic and `Krb5LoginModule` messages shed some lights when got stuck (though this gets better with experience).

Considering these if no objection I would turn debug on with the mentioned global flag.

SparkQA commented:
**[Test build #104322 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104322/testReport)** for PR 24204 at commit [`b3c467b`](https://github.com/apache/spark/commit/b3c467b384c1c56e73a390fc12c87350def99dd4).

SparkQA commented:
**[Test build #104322 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104322/testReport)** for PR 24204 at commit [`b3c467b`](https://github.com/apache/spark/commit/b3c467b384c1c56e73a390fc12c87350def99dd4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104322/
Test PASSed.

vanzin commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9850/
Test PASSed.

SparkQA commented:
**[Test build #104523 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104523/testReport)** for PR 24204 at commit [`b3c467b`](https://github.com/apache/spark/commit/b3c467b384c1c56e73a390fc12c87350def99dd4).

SparkQA commented:
**[Test build #104523 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104523/testReport)** for PR 24204 at commit [`b3c467b`](https://github.com/apache/spark/commit/b3c467b384c1c56e73a390fc12c87350def99dd4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104523/
Test PASSed.

vanzin commented:
Merging to master.
